feat(indev): add multi touch gestures

etag4048 opened the pull request:
A clear and concise description of what the bug or new feature is.

@kisvegabor 
This a draft of the new multi-touch recognition feature

etag4048 commented:
Related issue: https://github.com/lvgl/lvgl/issues/6014

kisvegabor commented on src/indev/lv_indev_gesture.h:
```diff
@@ -0,0 +1,115 @@
+	lv_point_t point;		/* Coordinates of the touch */
+	lv_indev_state_t state;		/* The state i.e PRESSED or RELEASED */
+	uint8_t id;			/* Identification/slot of the contact point */
+	uint32_t ts;			/* Timestamp in milliseconds */
```
Avoid unclear abbreviations
```suggestion
	uint32_t timestamp;			/* Timestamp in milliseconds */
```

kisvegabor commented on src/indev/lv_indev_gesture.h:
```diff
@@ -0,0 +1,115 @@
+	lv_indev_gesture_type_t type;    
+
+	/* Motion descriptor, stores the coordinates and velocity of a contact point */
+	lv_indev_gesture_motion_t motions[LV_GESTURE_MAX_TOUCHES];
```
```suggestion
	lv_indev_gesture_motion_t motions[LV_GESTURE_MAX_POINTS];
```

kisvegabor commented on src/indev/lv_indev_gesture.h:
```diff
@@ -0,0 +1,115 @@
+typedef struct {
+	lv_indev_gesture_state_t state;
+	lv_indev_gesture_t *info;
+	float scale;
```
Avoid floats by defaults
```suggestion
	uint32_t scale; /*256: no scale, 128: half size, 512 double size*/
```

kisvegabor commented on src/indev/lv_indev_gesture.h:
```diff
@@ -0,0 +1,115 @@
+	lv_indev_gesture_state_t state;
+	lv_indev_gesture_t *info;
+	float scale;
+	float rotation;
```
```suggestion
	int32_t rotation; /*0.1 degree resolution. E.g. 456 means 45.6 degree*/
```

kisvegabor commented on src/indev/lv_indev_gesture.h:
```diff
@@ -0,0 +1,115 @@
+ * @param touches 		pointer to the first element of the collected touch events
+ * @param touch_cnt 		length of passed touch event array.
+ */
+void lv_indev_gesture_detect_pinch(lv_indev_gesture_pinch_t *gesture, lv_indev_touch_data_t *touches, uint16_t touch_cnt);
```
If only `lv_indev_gesture_pinch_t` and `lv_indev_touch_data_t` are needed on the API level let's create an `lv_indev_gesture_private.h` for the other types and include it only in the C file. 

Or add the non API types in the C file.

kisvegabor commented:
I couldn't test it yet, but I checked the API in detail. I have a few smaller refactoring ideas to better follow LVGL naming conventions, but nothing major.

However, there is a larger thing: I wonder if we really need the concept of gesture recognizer in `lv_indev_t`. Note that our internal gesture recognizer algorithms are optional. It can happen that the driver already provides the gesture information, so we should provide only the high level information to the indev and keep the internal states hidden. So my notes:
- In `lv_indev_data_t` we can set an `lv_indev_gesture_type_t gesture_type` field and a `void * gesture_data`. `gesture_data` is interpreted according to the type and can be get in the event.  
- The gesture recognizer is really needed for `lv_indev_gesture_detect_pinch` to keep track of its state but it's used only in the driver.
- `lv_indev_gesture_state_t` can be tracked by LVGL as it knows when the gesture was set for the first time. 
- In `lv_indev_data_t` set `point.x/y` to the central point in the driver (if any) or to the primary point. It'd be more explicit.
- We should also compensate the scale thresholds. The user shouldn't see these internals. Instead when gesture becomes active (regardless of the threshold) we should start count the scale from there. So maybe internally we start recognizing the pinch only from 125% scale, but from the user's point of it's the 100% and 126% is the 101%.   
- I'm thinking about if we really need events for each gesture type, or a single `LV_EVENT_GESTURE` (where we can get the gesture type) could be enough. This way we can easily add `GESTURE_START/END` events as well, plus it could be backward compatible with the current gesture event.  

Ideally the scale event of the demo could be this simple:
```c
static void label_scale(lv_event_t *e)
{
   float scale = lv_event_get_pintch_scale(e); //If it's not a PINCH event return 1.0
   if(scale < 0.3) scale = 0.3;

    lv_point_t p;
    lv_indev_get_point(lv_indev_active(), &p);

    label_width = RECT_INIT_WIDTH * scale;
    label_height = RECT_INIT_WIDTH * scale;
    label_x = p.x - label_width / 2;
    label_y = p.y - label_height / 2;

    LV_LOG_TRACE("label scale: %g label x: %g label y: %g w: %g h: %g\n",
            scale, label_x, label_y, label_width, label_height);

    /* Update position and size */
    lv_style_set_width(&label_style, (int)label_width);
    lv_style_set_height(&label_style, (int)label_height);
    lv_style_set_x(&label_style, (int)label_x);
    lv_style_set_y(&label_style, (int)label_y);
}
```

What do you think?

etag4048 commented:
@kisvegabor Thanks for the review, I agree about the separating the indev and the gesture recognizer and will refactor accordingly.

>     * I'm thinking about if we really need events for each gesture type, or a single `LV_EVENT_GESTURE` (where we can get the gesture type) could be enough. This way we can easily add `GESTURE_START/END` events as well, plus it could be backward compatible with the current gesture event.

I think it is useful to have an event per gesture, only for fact that the user can set an event gesture event callback of a specific type, on a specific object. Having a general callback for all gesture types would require the user to always check the type of the gesture.

Thanks for the help with sample program, fetching the gesture info and style updating is much cleaner in your proposal, however the original demo supported subsequent zooms. By using a static variable within the scope of the function, that was used to save the width/height of the rectangle just when the gesture was recognized, the calculation is performed on this initial value during the pinch gesture. At the end in the _ENDED branch it was set back to the sentinel values, ready for the next run. Using a constant means if the rectangle is zoomed once when it's zoomed again it won't take into account that it was already scaled previously (and fetching the width/height of lv_obj_t, results in it being scaled way too much).

kisvegabor commented:
> I think it is useful to have an event per gesture, only for fact that the user can set an event gesture event callback of a specific type, on a specific object.

I see and it makes sense. So we need to choose between 
1. compatibility with the current simple gesture processing (up/down/left/right). (`LV_EVENT_GESTURE`)
2. saving a some code for the user. (`LV_EVENT_GESTURE_PINCH/ROTATE/etc`)

I still believe we should choose 1) because backward compatibility and consistency is higher priority than saving 2-3 lines of code. Having `GESTURE_BEGIN/END` would be great too as it's similar to `SCROLL_BEGIN/END`.

So let's go for 1) now. 

> however the original demo supported subsequent zooms.

I see, however it's an example to show the major concept. We shouldn't overwhelm the users by shown too much. :slightly_smiling_face:

etag4048 commented:
@kisvegabor No problems, let's go with a single gesture event for all. My plan included doing a nice demo program to demonstrate capabilities of this feature and also test every kind of gestures to see if the API works, and keeping the short snippet for the docs ? Is it ok for you ?

kisvegabor commented:
Sounds good! Typically the demos are something that can be shown on a fair. So just try to make the demo interesting and ascetic. :slightly_smiling_face:

etag4048 commented:
@kisvegabor I've refactored the code to use `gesture_type` and `gesture_data` instead of the previous approach. However, I don't see how one would dispatch GESTURE_END/START if we don't have access to the state of the gesture inside of `lv_indev.c` I you have an idea please let me know. I we can obtain the state of the gesture via a function call to `lv_event_get_gesture_state`

etag4048 commented:
@kisvegabor Ok will refactor accordingly, will be ready on monday.

etag4048 commented:
@kisvegabor CI is fixed, I've also moved the demo to examples instead, I've started to create a private header. however I believe it's better to separate detection logic from the generic API once we removed the reliance on float types. Because the main structure uses floats.

lhdjply commented:
https://github.com/lvgl/lvgl/blob/a5dce5c48e48a7b3a680d82766e5385dd570cb24/examples/porting/lv_port_indev_template.c#L179-L215

Can you modify the `lv_port_indev_template.c` to support  multi touch gestures

kisvegabor reviewed (approved):
Thank you Erik! 

As we discussed during our call still there are few improvement opportunities:
- add more gesture detectors besides pinch: rotate, swipe
- API fine tuning
- evaluate integer only logic (low priority)

Anyway, I think it's a great start and complete enough to merge it. 

@liamHowatt please review it.

kisvegabor commented:
> Can you modify the lv_port_indev_template.c to support multi touch gestures

Good idea! @etag4048 is on a holiday now, and when he is back a follow up PR will be added to improve a few things. Let's handle it there as well.

liamHowatt commented on src/indev/lv_indev_private.h:
```diff
@@ -112,6 +112,9 @@ struct _lv_indev_t {
     lv_anim_t * scroll_throw_anim;
+
+    lv_indev_gesture_type_t gesture_type;
+    void * gesture_data;
```
Probably this could be `lv_indev_gesture_recognizer_t *` since that's the only thing assigned to it but it's ok.

liamHowatt commented on src/indev/lv_indev_gesture_private.h:
```diff
@@ -0,0 +1,93 @@
+ *********************/
+#include "../core/lv_obj.h"
+
+#if LV_USE_GESTURE_RECOGNITION
+
+/*********************
```
Should include lv_indev_gesture.h

liamHowatt commented on src/drivers/wayland/lv_wayland.c:
```diff
@@ -117,11 +116,12 @@ struct input {
-        lv_indev_state_t state;
-    } touch;
+#if LV_USE_GESTURE_RECOGNITION
+    lv_indev_touch_data_t touches[10];
+    uint8_t touch_event_cnt;
+    uint8_t primary_id;
+    lv_indev_gesture_recognizer_t recognizer;
+#endif
```
Does Wayland normal touch still work with `LV_USE_GESTURE_RECOGNITION` disabled? Or does Wayland depend on it being enabled? If it's required, then all the guards could go away and a single `#if !LV_USE_GESTURE_RECOGNITION ... #error ... #endif` could go at the top.

liamHowatt reviewed (approved):
High level it looks good. We can address the things I mentioned later.
